src/lang/std/engineConnection.ts

```diff
@@ -2121,18 +2121,28 @@ export class EngineCommandManager extends EventTarget {
    * When an execution takes place we want to wait until we've got replies for all of the commands
    * When this is done when we build the artifact map synchronously.
    */
-  async waitForAllCommands() {
+  async waitForAllCommands(useFakeExecutor = false) {
     await Promise.all(Object.values(this.pendingCommands).map((a) => a.promise))
-    this.artifactGraph = createArtifactGraph({
-      orderedCommands: this.orderedCommands,
-      responseMap: this.responseMap,
-      ast: this.getAst(),
+    setTimeout(() => {
+      // the ast is wrong without this one tick timeout.
+      // an example is `Solids should be select and deletable` e2e test will fail
+      // because the out of date ast messes with selections
+      this.artifactGraph = createArtifactGraph({
+        orderedCommands: this.orderedCommands,
+        responseMap: this.responseMap,
+        ast: this.getAst(),
+      })
+      if (useFakeExecutor) {
+        // mock executions don't product an artifactGraph, so this will always be empty
+        // skipping the below logic to wait for the next real execution
+        return
+      }
+      if (this.artifactGraph.size) {
+        this.deferredArtifactEmptied(null)
+      } else {
+        this.deferredArtifactPopulated(null)
+      }
```

Comment:
Nice find! Could we add `TODO: race condition` to make the comment brutally clear?

Reply:
Shouldn't this dependency be handled within the react code that is causing the problem instead of this function itself?

Typically when I wrote vue code you could do something like

```
await Vue.nextTick()
runYourFunctionSinceTheInternalsOfVueAndStateAreUpdated()
```

Instead of baking that into the function itself, the function isn't the problem.

Reply:
So I don't think this has anything to do with react, the race is at the end of a execution no?

I do like `await nextTick(); // do stuff` but it didn't work, I have no idea why, made no sense to me.

If anyone wants to play with it, the test that fails when removing time out is `Solids should be select and deletable`

I at least added Franks race comment.